### ISSUE TITLE
Types don't work as expected

### ISSUE THREAD

**Issue by mendrik** [microsoft/TypeScript#46386]
# Bug Report

So I have a semi-complex [piece of code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAIghsOUC8UDeBYAUFKcDmEAXFAHYCuAtgEYQBO2upclxUAzsHQJan7YBfbNlCQoABToQAJtwDGCCAB4AKgD4U6KHID2AGxIBrCCB0AzKCoDcUAG5wDlgNqSZ8xarVOA5Lr3eAXQCoISxRaEkdMABRAEdyB3ZNJQApKAgAD2AIUmkk+EQAGigAaXSsnLyoY1MLFLUACj8qUhIS4vs9cjYUpxKAgEoUDVdZBWzUtWEsXVJOKDA6KLiEvXYSSJj4xM0m-Q6HIeQNBrRtfbsHEIHp8KgAeTppVXLs3PyEOA1UBp1qACsSCojhpbDpuNIROAItxIABGZIADVelQ+RVKKPe1RM5igiMafjaB0ciKc4VxfkGwweTyU+KhYnEsIgACYkZiqgUvg1FjINlIxh58SCac96WFoRJmQBmdmZN6cz7FMry1HY2p4gn6IlQaQQdhyEjUHT6CBwUgix5iqZYbCzeZgOEbZkIn5+Ykiv7-algiFQAAMdp0c2ACxZzsgbNQvOk1K9PvBscDM2DDulEYgsrdFz1Bs9AITfuT2AA9CWoMAABbQfAJJ5QbhJf7keZIaRUSggCs6Bt6LqcOiKCvVhZLah6CCUYrsHvSYPeUPscj4QjzFu8fB4KBOc0gYq7-ekEDBYDkMAThkwyAAcTrzwAggAlR8cpJSOBz0h6Lu7pwBb5QE+j4+BOfBVoEr5QLKAD8gHPlAJCkBAtj0LckoqPqoaoC8qpYlyxTiJBVqeH+jTSJ8QIig0AB0tFgMy6wSCKvqQlgZYVuhmHJCokFcmRFGWCKSj3CqFRYkyN53sJtLqKRjS0dR9GQIxInMYmQYhhWmFAlxyCMLqnzDPpNF0QxIL6bgLEBngSS7rcmGnHghAkCyxTMKwJDeOa0jvuw3jXDycINF5hDeMULIDMUYAsjySxbKs7DBe5EBhVA3jSNI3gDJFCzSklLApcUXDdNl2BAA) set up that works just fine, as can be seen in the last line which doesn't throw any errors.

However I want to be able to add an additional guard based on the argument `...pipe` and I don't know how or even where to do it. You can see the problem when the second line with `type Test` is commented in instead. Somehow everything falls apart.

I would expect the "array.length" guard to pass just fine or resort to never if we use anything but 3 pipes as arguments.

### 🔎 Search Terms

None.

### 🕗 Version & Regression Information

4.4.3,  probably never worked.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=30&ssc=64&pln=30&pc=68#code/C4TwDgpgBAIghsOUC8UDeBYAUFKcDmEAXFAHYCuAtgEYQBO2upclxUAzsHQJan7YBfbNlCQoABToQAJtwDGCCAB4AKgD4U6KHID2AGxIBrCCB0AzKCoDcUAG5wDlgNqSZ8xarVOA5Lr3eAXQCoISxRaEkdMABRAEdyB3ZNJQApKAgAD2AIUmkk+EQAGigAaXSsnLyoY1MLFLUACj8qUhIS4vs9cjYUpxKAgEoUDVdZBWzUtWEsXVJOKDA6KLiEvXYSSJj4xM0m-Q6HIeQNBrRtfbsHEIHp8KgAeTppVXLs3PyEOA1UBp1qACsSCojhpbDpuNIROAItxIABGZIADVelQ+RVKKPe1RM5igiMafjaB0ciKc4VxfkGwweTyU+KhYnEsIgACYkZiqgUvg1FjINlIxh58SCac96WFoRJmQBmdmZN6cz7FMry1HY2p4gn6IlQaQQdhyEjUHT6CBwUgix5iqZYbCzeZgOEbZkIn5+Ykiv7-algiFQAAMdp0c2ACxZzsgbNQvOk1K9PvBscDM2DDulEYgsrdFz1Bs9AITfuT2AA9CWoMAABbQfAJJ5QbhJf7keZIaRUSggCs6Bt6LqcOiKCvVhZLah6CCUYrsHvSYPeUPscj4QjzFu8fB4KBOc0gYq7-ekEDBYDkMAThkwyAAcTrzwAggAlR8cpJSOBz0h6Lu7pwBb5QE+j4+BOfBVoEr5QLKAD8gHPlAJCkBAtj0NMZYVpKKj6qGqAvKqWJcsU4iQVanh-o00ifECIoNAAdPRYDMusEgir6kISmIWHzLhKiQVyFFUZYIpKPcKoVFiTI3neIm0uo5GNPRtGMZAzGiaxiZBiGFbYUC2EoIwuqfMMBl0QxTEggZuBsQGeBJLutzYaceCECQLLFMwrAkN45rSO+7DeNcPJwg03mEN4xQsgMxRgCyPJLFsqzsCFHkQOFUDeNI0jeAMUULNKyUsKlxRcN0OXYEAA)

### 💻 Code

```ts
type Data = {
  age: number
  name: string
}

type Predicate<T> = { col: keyof T; val: T[Predicate<T>['col']] }
type PropEquals = <J extends Data, K extends keyof J>(column: K, value: J[K]) => Predicate<J>

const propEquals: PropEquals = (col, val) => ({ col, val })

type Ord<T extends Data> = (obj: T) => void
type Pipe1 = <X extends Data, K extends keyof X>(col: K, val: X[typeof col]) => Ord<X>
type Pipe2 = <X extends Data>(pred: Predicate<X>) => Ord<X>
type Pipe3 = <X extends Data, K extends keyof X>(col: K, desc: boolean) => Ord<X>

const p1: Pipe1 = (col, val) => obj => void 0
const p2: Pipe2 = pred => obj => void 0
const p3: Pipe3 = (col, desc) => obj => void 0

// the guard is just a dummy to illustrate the problem, so don't suggest using a [any, any, any] tuple
type PipeGuard<ARR extends readonly any[]> = ARR['length'] extends 3 ? ARR : never

// type Test = <T extends Data, P extends Ord<T>[]>(data: T) => (...pipes: P) => void
type Test = <T extends Data>(data: T) => <OK extends PipeGuard<Ord<T>[]>>(...pipes: OK) => void
const test: Test =
  data =>
  (...pipes) =>
    void 0 as any

test({ age: 2, name: 'andreas' })(p1('age', 2), p2(propEquals('name', 'dd')), p3('name', true))
```

### 🙁 Actual behavior

False error report.

### 🙂 Expected behavior

last line should pass with 2nd `type Test` commented in.

**Comment by mendrik** [comment]
seems all is working as it should. closing. thanks @orouz
